### PR TITLE
Dependency injection

### PR DESCRIPTION
**Pull request by romppane**
Made lots of refactoring structurally and visually.
Renamed lots of files due to either their role changing or because of my bad naming habits from the past.
Introduced Dependency injection. which had a big impact on the overall structure of the project:
Server.ts got a lot more readable, routers were split to controllers and services improving the single responcibility idea when compared to having just a router. 
Middleware is also now being injected to the classes instead of hard importing.

### CODE REVIEW

**Comment by henrikra** [comment]
Please fix the conflict 💥

File: package.json [hunk 1]
```diff
@@ -1,53 +1,54 @@
...
+    "moment-timezone": "^0.5.26",
+    "morgan": "^1.9.1",
+    "mysql": "^2.14.1",
+    "reflect-metadata": "^0.1.10",
+    "ts-node-dev": "^1.0.0-pre.40",
+    "typeorm": "^0.2.18",
+    "typescript": "^3.5.3"
+  },
+  "devDependencies": {
+    "@types/body-parser": "^1.17.0",
```
Comment (henrikra) [inline package.json]: Do you still need this? :D

File: package.json [hunk 2]
```diff
@@ -1,53 +1,54 @@
...
+    "@types/boom": "^7.2.1",
+    "@types/express": "^4.17.0",
+    "@types/glob": "^7.1.1",
```
Comment (henrikra) [inline package.json]: what is this used for? 🤔

File: server/controllers/post-controller.ts [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+    this.postService = options.postService;
+    this.userService = options.userService;
+    this.validatePost = options.validatePost;
+    this.validatePostPUT = options.validatePostPUT;
+    this.validateParams = options.validateParams;
+    this.initializeRoutes();
+  }
+
+  initializeRoutes() {
+    this.router.get('/', this.getAll.bind(this));
```
Comment (henrikra) [inline server/controllers/post-controller.ts]: I would use fat arrow functions to be able to get rid of binding. But you can choose which is better in your opinion 🤷‍♂

File: server/controllers/post-controller.ts [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+      } else {
+        next(this.notFound);
+      }
+    } catch (error) {
+      next(Boom.badImplementation());
+    }
+  }
+
+  // Create post
+  async post(req: Request, res: Response, next: NextFunction) {
```
Comment (henrikra) [inline server/controllers/post-controller.ts]: Would `create` be better name for this function? In general if there has to be comment to clarify the function then the function is badly named :P

File: server/controllers/post-controller.ts [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+      const post = await this.postService.save(
+        req.body.ownerId,
+        req.body.category,
+        req.body.message,
+        this.userService
+      );
+      if (post) {
+        res.status(201).send(post);
+      } else {
+        next(Boom.notFound("User doesn't exist"));
```
Comment (henrikra) [inline server/controllers/post-controller.ts]: Hmm I am not sure if 404 is good http error code when creation of something failed ;)

File: server/di-container.ts [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+  validateReplyPUT
+} from './middleware/validation';
+
+const container = awilix.createContainer({
+  injectionMode: awilix.InjectionMode.PROXY
+});
+
+export function configureContainer() {
+  // Add exception handler
+  // Start server only if connection established.
```
Comment (henrikra) [inline server/di-container.ts]: Hmm these comment don't make much sense here :D

File: server/service/post-service.ts [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+      return undefined;
+    }
+  }
+
+  delete(id: number) {
+    return this.repository.delete(id);
+  }
+
+  update(id: number, obj: QueryDeepPartialEntity<Post>) {
+    this.repository.update(id, obj);
```
Comment (henrikra) [inline server/service/post-service.ts]: Hmm isn't this async operation that we have to wait before returning? Also can you check what `update` returns? Could it be the updated object?

File: server/service/reply-service.ts [hunk 8]
```diff
@@ -0,0 +1,52 @@
+import { Reply } from '../entities/reply';
+import { Dependencies } from '../types';
+import { QueryDeepPartialEntity } from 'typeorm/query-builder/QueryPartialEntity';
+
+export class ReplyService {
+  // Can't use Repository<Entity>, needs specific?
+  // Should the service know all the repositories or get access to services via parameters from controllers?
+  repositories: any;
```
Comment (henrikra) [inline server/service/reply-service.ts]: Just add all needed repos individually 👍

File: server/controllers/post-controller.ts (hunk 4 above)
Comment (romppane) [reply server/controllers/post-controller.ts]: At the time it was serving under the name of router, it seemed suitable for the methods to be named after the HTTP Requests! I agree with the renaming though! Switched gets to reads as well!

File: server/controllers/post-controller.ts (hunk 5 above)
Comment (romppane) [reply server/controllers/post-controller.ts]: It's a tough one! The request goes through validation before it reaches the service layer and in the service layer there isn't much more left other than making and returning the post or failing and returning nothing which would pretty much mean that the user doesn't indeed exist!

File: server/service/post-service.ts (hunk 7 above)
Comment (romppane) [reply server/service/post-service.ts]: This we went through in tv and I checked the updateResult object which was just what I expected! Fixed all the update methods in to using .then though!

**Comment by henrikra** [comment]
Looks good 👍  Approved!

**Pull request merged**